isis/src/galileo/apps/gllssical/gllssical.cpp

```diff
@@ -0,0 +1,535 @@
+/** This is free and unencumbered software released into the public domain.
+
+The authors of ISIS do not claim copyright on the contents of this file.
+For more details about the LICENSE terms and the AUTHORS, you will
+find files of those names at the top level of this repository. **/
+
+/* SPDX-License-Identifier: CC0-1.0 */
+
+#include "ProcessByLine.h"
+#include "Buffer.h"
+#include "Camera.h"
+#include "iTime.h"
+#include "SpecialPixel.h"
+#include "Spice.h"
+#include "TextFile.h"
+#include "NaifStatus.h"
+
+#include "gllssical.h"
+
+using namespace std;
+
+/**
+ * Shutter file:
+ * The shutter on the galileo SSI camera took from
+ * about 1ms at the top of the camera to about 1.5ms
+ * at the bottom due to friction. The shutter offset file in
+ * isis 2 is rotated 90 degrees.
+ *
+ * Some of the values and equations in the program were verified
+ *   using the book "In Orbit at Jupiter, Contributions of the Galileo
+ *   Science Team," Section H Part I.
+ */
+
+
+namespace Isis {
+  
+  // Globals
+  static vector<double> weight;
+  static double scaleFactor0; // Entire scale factor that e is multiplied by
+  static double scaleFactor; // A1 or A2 in the equation, depending on units
+  static double dcScaleFactor;
+  static double exposureDuration;
+  static bool eightBitDarkCube; // Is the dark cube of type unsigned byte?
+  static bool iof; // Determines output units (i.e. I/F or radiance)
+  static double s1;
+  static double s2;
+  static double rsun;
+  static double cubeConversion;
+  static double gainConversion;
+  
+  static void Calibrate(vector<Buffer *> &in, vector<Buffer *> &out);
+  static FileName FindDarkFile(Cube *icube);
+  static FileName FindGainFile(Cube *icube);
+  static FileName FindShutterFile(Cube *icube);
+  static FileName ReadWeightTable(Cube *icube);
+  static FileName GetScaleFactorFile();
+  static int getGainModeID(Cube *icube);
+  static void calculateScaleFactor0(Cube *icube, Cube *gaincube);
+  
+  void gllssical(UserInterface &ui, Pvl *log) {
+    Cube icube(ui.GetFileName("FROM"));
```

Comment:
Technically speaking, _**ALL**_ input cubes can have attributes. Mission specific apps (like this one) are an odd case where it doesn't make sense to allow the user to mess with the bands, but we do allow it on the command line and gui, so it should be passed along and the app should throw if appropriate.